validator_client/slashing_protection/src/slashing_protection.rs

```diff
@@ -0,0 +1,402 @@
+use crate::attester_slashings::{check_for_attester_slashing, SignedAttestation};
+use crate::enums::{NotSafe, ValidityReason};
+use crate::proposer_slashings::{check_for_proposer_slashing, SignedBlock};
+use fs2::FileExt;
+use ssz::{Decode, Encode};
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Result as IOResult, Seek, SeekFrom, Write};
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+use types::{AttestationData, BeaconBlockHeader};
+
+/// Trait used to know if type T can be checked for slashing safety
+pub trait SafeFromSlashing<T> {
+    type U;
+
+    /// Checks if incoming_data is free from slashings, and if so updates the in-memory and writes it to the history file.
+    /// If the error returned is an IOError, do not sign nor broadcast the attestation.
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe>;
+}
+
+impl SafeFromSlashing<SignedAttestation> for HistoryInfo<SignedAttestation> {
+    type U = AttestationData;
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = check_for_attester_slashing(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                ValidityReason::SameVote => Ok(()),
+                _ => self
+                    .insert_and_write(SignedAttestation::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+impl SafeFromSlashing<SignedBlock> for HistoryInfo<SignedBlock> {
+    type U = BeaconBlockHeader;
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = check_for_proposer_slashing(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                // Casting the same vote, no need to add it to the history
+                ValidityReason::SameVote => Ok(()),
+                // New attestation, add it to memory and file history
+                _ => self
+                    .insert_and_write(SignedBlock::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+/// Struct used for checking if attestations or blockheaders are safe from slashing.
+#[derive(Debug)]
+pub struct HistoryInfo<T> {
+    file: File,
+    data: Vec<T>,
+}
+
+impl<T: Clone + Encode + Decode> HistoryInfo<T> {
+    /// Inserts the incoming data in the in-memory history, and writes it to the history file.
+    fn insert_and_write(&mut self, data: T, index: usize) -> IOResult<()> {
+        // Insert the incoming data in memory
+        self.data.insert(index, data);
+
+        // Writing new history to file
+        self.file.seek(SeekFrom::Start(0))?;
+        self.file.write_all(&self.data.as_ssz_bytes())?;
```

Comment:
We do have the risk here that if you crash during this operation you may corrupt the history.

There are two ways to deal with this:

1. Roll our own rollback system (e.g., write to a temp file, then copy it over the existing main file).
1. Instead of over-writing your history file, just append to it. This will make the on-disk file unordered, however you could just order it when you put it into memory.
1. Use an existing flat-file, low-throughput database implementation (e.g., SQLite).

SQLite already deals with journals, rollbacks, etc. It would also make writes something less than `O(v)` where `v` is all votes. It will have some space and time overhead, but given the history file is already pretty small, I think that's ok.

I probably should have realized this earlier, but I think it's worth having a look at using [`rusqlite`](https://docs.rs/rusqlite/0.20.0/rusqlite/). If it can be implemented easily (say, ~3 days) and doesn't have any show-stopping problems, this might be the best way to go.

For reference, checkout out [this doc](https://www.sqlite.org/howtocorrupt.html) on "how to corrupt an sqlite database file".

Reply:
Done!